File: src/stores/savedQueries.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'
import { invoke } from '@tauri-apps/api/core'
import type { SavedQuery } from '@/types/savedQuery'

export const useSavedQueriesStore = defineStore('savedQueries', () => {
  const queries = ref<SavedQuery[]>([])

  async function fetch() {
    queries.value = await invoke<SavedQuery[]>('get_saved_queries')
  }

  async function upsert(
    data: Omit<SavedQuery, 'id' | 'created_at' | 'updated_at'> & { id?: string; created_at?: string },
  ): Promise<string> {
    const now = new Date().toISOString()
    const existing = data.id ? queries.value.find(q => q.id === data.id) : null
    const entry: SavedQuery = {
      id: data.id ?? crypto.randomUUID(),
      created_at: existing?.created_at ?? data.created_at ?? now,
      updated_at: now,
      name: data.name,
      description: data.description,
      sql: data.sql,
      connection_id: data.connection_id,
      database: data.database,
    }
    await invoke('upsert_saved_query', { query: entry })
    await fetch()
    return entry.id
  }

  async function remove(id: string) {
    await invoke('delete_saved_query', { id })
    queries.value = queries.value.filter(q => q.id !== id)
  }

  return { queries, fetch, upsert, remove }
})

File: src-tauri/src/saved_queries.rs
use serde::{Deserialize, Serialize};
use tauri::State;
use uuid::Uuid;
use crate::state::AppState;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct SavedQuery {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub sql: String,
    pub connection_id: Option<Uuid>,
    pub database: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[tauri::command]
pub async fn get_saved_queries(state: State<'_, AppState>) -> Result<Vec<SavedQuery>, String> {
    let queries = state.saved_queries.read();
    let mut list: Vec<SavedQuery> = queries.values().cloned().collect();
    list.sort_by(|a, b| b.created_at.cmp(&a.created_at));
    Ok(list)
}

#[tauri::command]
pub async fn upsert_saved_query(state: State<'_, AppState>, query: SavedQuery) -> Result<(), String> {
    {
        let mut queries = state.saved_queries.write();
        queries.insert(query.id, query);
    }
    state.save_queries()
}

#[tauri::command]
pub async fn delete_saved_query(state: State<'_, AppState>, id: Uuid) -> Result<(), String> {
    {
        let mut queries = state.saved_queries.write();
        queries.remove(&id);
    }
    state.save_queries()
}
